### DISCUSSION TITLE
IEC 61850

### DISCUSSION THREAD

**Discussion by chrisdutz** [apache/plc4x#1016] [General]
Hi folks,

anyone ever come across this protocol? Seems to be a TCP-based fieldbus protocol.

Chris

**Comment by hutcheb** [comment]
Yes, it is very popular to connect devices in electrical substations. Timestamping is very important. I've haven't had a lot to do with it otherwise.

**Comment by chrisdutz** [comment]
Yikes ... the full spec only seems to kost 17k€ ;-) 
(https://webstore.iec.ch/publication/6028)

**Reply by chrisdutz** [reply]
But there seems to be an Apache 2.0 licensed implementation from some Fraunhofer guys: https://www.beanit.com/iec-61850/

**Reply by hutcheb** [reply]
That was the Swiss site? Not sure if you can buy from a separate country, www.evs.ee, but some are better than others.

**Reply by splatch** [reply]
You do not need to buy stuff from iec.ch, international norms can be bought from your local standardization body valid for Germany. Often these norms are re-published with additional cover as "local" norms i.e. in https://sklep.pkn.pl you can buy `PN-EN IEC 62351-..` for 40€/part. You might look for `DE-EN IEC 62351-..` and see if it will be equally expensive. ;-) I come over openmuc/Fraunhoffer stuff earlier. Their libs work fine. However, making a contribution or a change in these libs is pretty hard -  they are either not meant to be extended or contact with earlier contributors is quite difficult - I know that from project which involved jmbus in 2018. Hope that 62351 is better since it was adopted by another entity!

**Reply by chrisdutz** [reply]
Hehe ... wasn't thinking of contributing to that project ... was thinking of using it as one source of information for creating a PLC4X driver.

**Comment by ottobackwards** [comment]
IEC 61850 is a set of 3 protocol specification:
- MMS  which does normal kinds of things over tcp
- SV which is streamlined for value retrieval (if I recall correctly)
- GOOSE which is a link layer high speed data transfer protocol

MMS protocol runs on a stack like:
TCP
TPKT
COTP
MMS

Like S7 does.  In fact they both run on port 102.

It is possible that some vendors implement it differently though.  ABB devices for example seem to leave out parts of the COTP stack.

There are zeek analyzers for MMS and GOOSE, and Wireshark support as well.
There is an openiec61850 (C) project as well. but that is dual gpl/commeral license.

**Reply by chrisdutz** [reply]
Why does this whole think (with the xml descriptors) just keep on screaming "PROFINET" in my brain?

**Comment by ottobackwards** [comment]
There is also a specification for several xml schema that can define a network of IED devices that use 61850 etc
